### ISSUE TITLE
Rule proposal: Forbid imports with explicit side-effects

### ISSUE THREAD

**Issue by jfmengels** [benmosher/eslint-plugin-import#529]
IMO, modules that have side-effect just from requiring them are really bad. I would like to see a rule that forbids importing/requiring modules that are not assigned, which implicitly means there is a side-effect or simply not used and therefore useless.

### Invalid

```js
import 'dosideeffect';
import './dosideeffect';
require('./dosideeffect');
```

### Valid

```js
import foo from 'bar';

// Explicit effect
import doEffect from './dosideeffect';
doEffect();
```

As far as rule name goes: `no-unnassigned-import`, `no-side-effect-import`, ...? I'm leaning toward the former at the moment, though it doesn't capture the rationale of it.

**Comment by benmosher** [comment]
What if the rule instead enforced that modules not _have_ side effects? (or _only_ side effects?)

The `import`-er may not have the option to refactor to an explicit effect, and if he/she does, then a `no-side-effects` rule would flag the import-_ed_ module instead of the import-_ing_ module.

**Comment by jfmengels** [comment]
I could see the value for both, but I agree that a `no-side-effects` rule would be better. Except I don't see how you'd go and do that... :confused:

**Comment by jfmengels** [comment]
I'd like to work on this soon-ish, especially as it's pretty easy to implement.

I still don't see how you can enforce having modules not create side-effects. That'd be awesome if it was possible though...

**Comment by benmosher** [comment]
Works for me. Name is still not obvious, but that doesn't need to hamper development of the rule, obviously.

I thought I had said this in this issue before, but `no-side-effects` could warn on the first line of any module with `import`s, but no `export`s. Not sure how to handle the entry point of applications in that case, though. And again. no need to hold up the simple version of this rule (flagging `imports` without names).

**Comment by benmosher** [comment]
Yeah, fair enough. I think your original proposal makes the most sense, at least right now.
